squint.c

```diff
@@ -0,0 +1,1611 @@
+      if (is_nop(*scan) && !is_const(scan)) {
+         *scan = NOP;
+      }
+   }
+}
+
+/**********************************************************/
+/****************** amacc use-def analysis ****************/
+/**********************************************************/
+
+/* The following bit masks are used to transfer 'active' register
+   slots between the def/use descriptors and the A32 opcodes.
+*/
+int activeRegMask[16] =
+{
+   0x00000000,
+   0x0000000f,
+   0x00000f00,
+   0x00000f0f,
+   0x0000f000,
+   0x0000f00f,
+   0x0000ff00,
+   0x0000ff0f,
+   0x000f0000,
+   0x000f000f,
+   0x000f0f00,
+   0x000f0f0f,
+   0x000ff000,
+   0x000ff00f,
+   0x000fff00,
+   0x000fff0f
+};
+
+/**********************************************/
+
+#define MAX_RENAME_REG 8
+
+int regtest(int inst, int opmask)
+{
+   return ((inst & opmask) == opmask);
+}
+
+/* funcBegin and funcEnd instructions guaranteed not to be NOP */
+void create_inst_info(int *instInfo, int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   int *rInfo = instInfo;
+   int Rn, Rd, Rs, Rm;
+   int inst, instMask, op;
+
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      int info = 0;
+
+      /* skip code that won't be transformed */
+      if (is_nop(*scan)) {
+         int *end = skip_nop(scan, 1);
+         while (scan < end) {
+            if (*scan == NOP13) {
+               *rInfo++ = RI_func; // R0 set in func
+            }
+            else
+               *rInfo++ = info;
+            ++scan;
+         }
+      }
+
+      /* determine info about current instruction */
+      inst = *scan;
+      instMask   = (inst >> 24) & 0x0e;
+      Rn         = (inst >> 16) & 0x0f;
+      Rd         = (inst >> 12) & 0x0f;
+      Rs         = (inst >>  8) & 0x0f;
+      Rm         =  inst        & 0x0f;
+
+      /* Make note of registers eligible for renaming */
+
+      if (instMask == 0x02) { /* ALU_immed */
+         op = (inst >> 21) & 0x0f;
+         if ((op != 0x0d) && (op != 0x0f)) {
+             info |= RI_RnAct;
+         }
+         if ((op & 0x0c) != 0x08) { // not tst instructions
+             info |= RI_RdAct | RI_RdDest;
+         }
+      }
+      else if (instMask == 0x00) { /* ALU_register */
+         if ((inst & 0xf0) == 0x90) { /* mul */
+            info |= RI_RnAct | RI_RsAct | RI_RmAct | RI_RnDest;
+         }
+         else { /* not mul */
+            op = (inst >> 21) & 0x0f;
+            if ((op != 0x0d) && (op != 0x0f)) {
+                info |= RI_RnAct;
+            }
+            if ((op & 0x0c) != 0x08) {
+                info |= RI_RdAct | RI_RdDest;
+            }
+            if ((inst & 0x90) == 0x10) {
+               info |= RI_RsAct;
+            }
+            info |= RI_RmAct;
+         }
+      }
+      else if (instMask == 0x04 ||
+               (instMask == 0x06 && (inst & (1<<4)) == 0)) { /* MEM op */
+         int MEM_mask = (inst >> 20) & 0xd7;
+         if ( (MEM_mask == 0x41) || /* pop */
```

Comment:
Maybe separate the `MEM_mask == 0x41` (pop) case from ldr/ldrb?  Currently, they only share Line 498 but introduce a nested comparison on Line 499-501.